File: frontend/src/types/campaign.ts
/**
 * TypeScript types for Campaign and Message Template management (Story 6.4).
 *
 * Broadcasting & Template SMS/SNS — Owner creates bulk outreach campaigns
 * to customer segments via email (with SMS/Zalo stubs).
 */

export type ChannelType = "email" | "sms" | "zalo";
export type CampaignStatus = "draft" | "scheduled" | "sending" | "sent" | "failed";
export type RecipientStatus = "pending" | "sent" | "failed" | "opened" | "clicked";
export type SegmentType =
  | "all_customers"
  | "hot_leads"
  | "warm_leads"
  | "cold_leads"
  | "voucher_holders";

// --------- Message Templates ---------

export interface MessageTemplate {
  id: string;
  name: string;
  channel: ChannelType;
  subject: string | null;
  body: string;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface TemplateFormData {
  name: string;
  channel: ChannelType;
  subject: string;
  body: string;
}

export interface TemplatePreviewData {
  subject: string | null;
  body: string;
}

// --------- Campaigns ---------

export interface Campaign {
  id: string;
  name: string;
  channel: ChannelType;
  template_id: string;
  template_name: string;
  segment: SegmentType;
  voucher_id: string | null;
  voucher_code: string | null;
  status: CampaignStatus;
  scheduled_at: string | null;
  sent_at: string | null;
  total_recipients: number;
  sent_count: number;
  failed_count: number;
  created_at: string;
  updated_at: string;
}

export interface CampaignFormData {
  name: string;
  channel: ChannelType;
  template_id: string;
  segment: SegmentType;
  voucher_id: string | null;
  scheduled_at: string | null;
}

// --------- Analytics ---------

export interface CampaignAnalytics {
  total_recipients: number;
  sent_count: number;
  failed_count: number;
  open_rate: number;
  click_rate: number;
  voucher_redemptions: number;
}

export interface CampaignsSummary {
  total_campaigns: number;
  sent_campaigns: number;
  avg_open_rate: number;
  total_messages_this_month: number;
}

export interface SegmentInfo {
  segment: SegmentType;
  label: string;
  recipient_count: number;
}

// --------- Recipients ---------

export interface CampaignRecipient {
  id: string;
  email: string | null;
  recipient_name: string | null;
  status: RecipientStatus;
  sent_at: string | null;
  error_message: string | null;
}

// --------- API Response Wrappers ---------

export interface CampaignListApiResponse {
  data: Campaign[];
  meta: {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
  };
}

export interface CampaignDetailApiResponse {
  data: Campaign;
}

export interface CampaignsSummaryApiResponse {
  data: CampaignsSummary;
}

export interface CampaignAnalyticsApiResponse {
  data: CampaignAnalytics;
}

export interface TemplateListApiResponse {
  data: MessageTemplate[];
  meta: { total: number };
}

export interface TemplateDetailApiResponse {
  data: MessageTemplate;
}

export interface TemplatePreviewApiResponse {
  data: TemplatePreviewData;
}

export interface SegmentListApiResponse {
  data: SegmentInfo[];
  meta: { total: number };
}

export interface RecipientListApiResponse {
  data: CampaignRecipient[];
  meta: {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
  };
}

// --------- UI Helpers ---------

export const SEGMENT_LABELS: Record<SegmentType, string> = {
  all_customers: "Tat ca khach hang",
  hot_leads: "Lead Hot",
  warm_leads: "Lead Warm",
  cold_leads: "Lead Cold",
  voucher_holders: "Nguoi so huu voucher",
};

export const CHANNEL_LABELS: Record<ChannelType, string> = {
  email: "Email",
  sms: "SMS",
  zalo: "Zalo OA",
};

export const STATUS_LABELS: Record<CampaignStatus, string> = {
  draft: "Ban nhap",
  scheduled: "Da lich",
  sending: "Dang gui",
  sent: "Da gui",
  failed: "That bai",
};

export const STATUS_COLORS: Record<CampaignStatus, string> = {
  draft: "text-stone-500 bg-stone-100",
  scheduled: "text-blue-700 bg-blue-100",
  sending: "text-amber-700 bg-amber-100",
  sent: "text-emerald-700 bg-emerald-100",
  failed: "text-red-700 bg-red-100",
};

File: backend/src/models/campaign.py
"""Pydantic schemas for Campaign and Message Template management (Story 6.4).

Broadcasting & Template SMS/SNS — Owner creates bulk outreach campaigns
with template messages sent to customer segments via email (with SMS/Zalo stubs).
"""

import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, field_validator


class ChannelType(str, Enum):
    """Supported broadcast channels. Email is functional; SMS/Zalo are stubs."""
    email = "email"
    sms = "sms"
    zalo = "zalo"


class CampaignStatus(str, Enum):
    """Campaign lifecycle statuses."""
    draft = "draft"
    scheduled = "scheduled"
    sending = "sending"
    sent = "sent"
    failed = "failed"


class RecipientStatus(str, Enum):
    """Per-recipient delivery statuses."""
    pending = "pending"
    sent = "sent"
    failed = "failed"


class SegmentType(str, Enum):
    """Audience segment types for campaign targeting."""
    all_customers = "all_customers"
    hot_leads = "hot_leads"
    warm_leads = "warm_leads"
    cold_leads = "cold_leads"
    voucher_holders = "voucher_holders"


# ---------- Message Template Schemas ----------

class TemplateCreateRequest(BaseModel):
    """Request to create a new message template."""
    name: str
    channel: ChannelType = ChannelType.email
    subject: Optional[str] = None       # required for email channel
    body: str

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Tên template không được để trống")
        if len(v) > 100:
            raise ValueError("Tên template tối đa 100 ký tự")
        return v

    @field_validator("body")
    @classmethod
    def body_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Nội dung template không được để trống")
        return v


class TemplateUpdateRequest(BaseModel):
    """Request to update an existing template. All fields optional."""
    name: Optional[str] = None
    channel: Optional[ChannelType] = None
    subject: Optional[str] = None
    body: Optional[str] = None

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Tên template không được để trống")
            if len(v) > 100:
                raise ValueError("Tên template tối đa 100 ký tự")
        return v

    @field_validator("body")
    @classmethod
    def body_not_empty(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Nội dung template không được để trống")
        return v


class TemplateResponse(BaseModel):
    """Response schema for a message template."""
    id: uuid.UUID
    name: str
    channel: ChannelType
    subject: Optional[str]
    body: str
    is_default: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TemplatePreviewRequest(BaseModel):
    """Request to preview a template with sample variable substitution."""
    sample_name: str = "Nguyen Van A"
    sample_voucher_code: str = "TETLUXV26"
    sample_voucher_value: str = "20%"
    sample_expiry_date: str = "31/12/2026"
    sample_shop_name: str = "Tailor Project"


class TemplatePreviewResponse(BaseModel):
    """Rendered preview of a template with sample data."""
    subject: Optional[str]
    body: str


# ---------- Campaign Schemas ----------

class CampaignCreateRequest(BaseModel):
    """Request to create a new broadcast campaign."""
    name: str
    channel: ChannelType = ChannelType.email
    template_id: uuid.UUID
    segment: SegmentType
    voucher_id: Optional[uuid.UUID] = None
    scheduled_at: Optional[datetime] = None    # null = immediate when send triggered

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Tên chiến dịch không được để trống")
        if len(v) > 200:
            raise ValueError("Tên chiến dịch tối đa 200 ký tự")
        return v


class CampaignUpdateRequest(BaseModel):
    """Request to update a draft campaign. Only draft campaigns can be updated."""
    name: Optional[str] = None
    channel: Optional[ChannelType] = None
    template_id: Optional[uuid.UUID] = None
    segment: Optional[SegmentType] = None
    voucher_id: Optional[uuid.UUID] = None
    scheduled_at: Optional[datetime] = None

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Tên chiến dịch không được để trống")
        return v


class CampaignResponse(BaseModel):
    """Response schema for a campaign with joined template/voucher info."""
    id: uuid.UUID
    name: str
    channel: ChannelType
    template_id: uuid.UUID
    template_name: str
    segment: SegmentType
    voucher_id: Optional[uuid.UUID]
    voucher_code: Optional[str]
    status: CampaignStatus
    scheduled_at: Optional[datetime]
    sent_at: Optional[datetime]
    total_recipients: int
    sent_count: int
    failed_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CampaignAnalytics(BaseModel):
    """Per-campaign delivery and engagement analytics."""
    total_recipients: int
    sent_count: int
    failed_count: int
    opened_count: int = 0   # 0 placeholder (requires email tracking pixel)
    clicked_count: int = 0  # 0 placeholder (requires tracked links)
    open_rate: float        # 0.0 placeholder (requires email tracking pixel)
    click_rate: float       # 0.0 placeholder (requires tracked links)
    voucher_redemptions: int


class CampaignsSummary(BaseModel):
    """Aggregate summary of all campaigns for a tenant."""
    total_campaigns: int
    sent_campaigns: int
    avg_open_rate: float
    total_messages_this_month: int


class SegmentInfo(BaseModel):
    """Segment option with its current recipient count."""
    segment: SegmentType
    label: str
    recipient_count: int


class CampaignRecipientResponse(BaseModel):
    """Per-recipient record in a campaign send log."""
    id: uuid.UUID
    email: Optional[str]
    recipient_name: Optional[str]
    status: RecipientStatus
    sent_at: Optional[datetime]
    error_message: Optional[str]

    model_config = {"from_attributes": True}
